File: dashboard/server/board/sources/lever.js
import { COMPANIES } from "./companies.js";

/**
 * Lever public postings API.
 *
 * `createdAt` is epoch milliseconds rather than ISO — the only shape
 * difference worth noting across the three ATSs.
 */

const API = "https://api.lever.co/v0/postings";

function normalize(job, slug) {
  const published = job.createdAt ? new Date(job.createdAt).toISOString() : null;
  return {
    posting_id: `lever:${slug}:${job.id}`,
    title: job.text,
    company: slug,
    location: job.categories?.location || null,
    job_url: job.applyUrl || job.hostedUrl,
    posted_at: published ? published.slice(0, 10) : null,
    posted_at_precise: published,
    applicants: null,
  };
}

export const leverSource = {
  name: "lever",

  async fetchRecent({ windowSeconds }) {
    const cutoff = Date.now() - windowSeconds * 1000;
    const out = [];

    for (const slug of COMPANIES.lever) {
      try {
        const res = await fetch(`${API}/${slug}?mode=json`);
        if (!res.ok) {
          console.warn(`[board] lever/${slug} -> HTTP ${res.status}, skipped`);
          continue;
        }
        const jobs = await res.json();

        for (const job of Array.isArray(jobs) ? jobs : []) {
          const p = normalize(job, slug);
          if (!p.posted_at_precise) continue;
          if (Date.parse(p.posted_at_precise) < cutoff) continue;
          out.push(p);
        }
      } catch (err) {
        console.warn(`[board] lever/${slug} failed: ${err.message}`);
      }
    }

    return out;
  },

  async enrich(posting) {
    return posting;
  },

  get sliceCount() {
    return COMPANIES.lever.length;
  },
};

File: dashboard/server/board/sources/lever.py
"""Lever public postings API.

`createdAt` is epoch milliseconds rather than ISO — the only shape
difference worth noting across the ATSs.
"""

import logging

from board.sources.companies import companies
from board.sources.http import client
from board.timeutil import iso_from_ms, now_ms, parse_ms

log = logging.getLogger("board")

API = "https://api.lever.co/v0/postings"


def _normalize(job: dict, slug: str) -> dict:
    published = iso_from_ms(job["createdAt"]) if job.get("createdAt") else None
    return {
        "posting_id": f"lever:{slug}:{job['id']}",
        "title": job.get("text"),
        "company": slug,
        "location": (job.get("categories") or {}).get("location"),
        "job_url": job.get("applyUrl") or job.get("hostedUrl"),
        "posted_at": published[:10] if published else None,
        "posted_at_precise": published,
        "applicants": None,
    }


class LeverSource:
    name = "lever"

    async def fetch_recent(self, window_seconds: int) -> list[dict]:
        cutoff = now_ms() - window_seconds * 1000
        out = []
        for slug in companies("lever"):
            try:
                res = await client.get(f"{API}/{slug}?mode=json")
                if res.status_code != 200:
                    log.warning("[board] lever/%s -> HTTP %d, skipped", slug, res.status_code)
                    continue
                jobs = res.json()
                for job in jobs if isinstance(jobs, list) else []:
                    p = _normalize(job, slug)
                    if not p["posted_at_precise"]:
                        continue
                    if parse_ms(p["posted_at_precise"]) < cutoff:
                        continue
                    out.append(p)
            except Exception as err:
                log.warning("[board] lever/%s failed: %s", slug, err)
        return out

    async def enrich(self, posting: dict) -> dict:
        return posting

    @property
    def slice_count(self) -> int:
        return len(companies("lever"))


lever_source = LeverSource()
